src/rt/minfo.d

```diff
@@ -53,6 +53,128 @@ struct ModuleGroup
         return _modules;
     }
 
+    // this function initializes the bookeeping necessary to create the
+    // cycle path, and then creates it. It is a precondition that src and
+    // target modules are involved in a cycle.
+    //
+    // The delegate is a helper to map module info pointers to index into the modules array
+    private int[] genCyclePath(int srcidx, int targetidx,
+            scope int delegate(immutable(ModuleInfo)*) findModule)
+    {
+        import core.bitop : bt, btc, bts;
+
+        // set up all the arrays. Use the GC, we are going to exit anyway.
+        int[] cyclePath;
+        int[] distance;
+        int[][] edges;
+        distance.length = _modules.length;
+        edges.length = _modules.length;
+        immutable nwords = (_modules.length + 8 * size_t.sizeof - 1) / (8 * size_t.sizeof);
+        immutable flagbytes = nwords * size_t.sizeof;
+        auto reachable = cast(size_t*) malloc(flagbytes);
+        scope (exit)
+            .free(reachable);
+
+        foreach (i, m; _modules)
+        {
+            // use bit array to prevent duplicates
+            // https://issues.dlang.org/show_bug.cgi?id=16208
+            memset(reachable, 0, flagbytes);
+            foreach (e; m.importedModules)
+            {
+                auto impidx = findModule(e);
+                if (impidx != -1 && impidx != i)
+                    if (!bts(reachable, impidx))
+                        edges[i] ~= impidx;
+            }
+        }
+
+        // determine the shortest path between two modules. Uses dijkstra
+        // without a priority queue. (we can be a bit slow here, in order to
+        // get a better printout).
+        void shortest(int start, int target)
```

Comment:
I didn't check hard enough to see if there are nested references. If there are none probably won't make a dent in the cost either way.